src/circuit/circuit_builder.rs

```diff
@@ -1101,7 +1116,58 @@ where
             self.connect_stream(input_stream1, id, input_preference1);
             self.connect_stream(input_stream2, id, input_preference2);
             (node, output_stream)
-        }))
+        })
+    }
+
+    /// Add a N-ary operator (see [`NaryOperator`]).
+    pub fn add_nary_operator<'a, I, O, Op, Iter>(
+        &'a self,
+        operator: Op,
+        input_streams: Iter,
+    ) -> Stream<Self, O>
+    where
+        I: Data,
+        O: Data,
+        Op: NaryOperator<I, O>,
+        Iter: IntoIterator<Item = &'a Stream<Self, I>>,
+    {
+        let pref = operator.input_preference();
+        self.add_nary_operator_with_preference(operator, input_streams, pref)
+    }
+
+    /// Like [`Self::add_nary_operator`], but overrides the ownership
+    /// preference with `input_preference`.
+    pub fn add_nary_operator_with_preference<'a, I, O, Op, Iter>(
+        &'a self,
+        operator: Op,
+        input_streams: Iter,
+        input_preference: OwnershipPreference,
+    ) -> Stream<Self, O>
+    where
+        I: Data,
+        O: Data,
+        Op: NaryOperator<I, O>,
+        Iter: IntoIterator<Item = &'a Stream<Self, I>>,
+    {
+        let input_streams: Vec<Stream<_, _>> = input_streams.into_iter().cloned().collect();
+        self.add_node(|id| {
+            self.log_circuit_event(&CircuitEvent::operator(
+                GlobalNodeId::child_of(self, id),
+                operator.name(),
+            ));
+
+            let node = NaryNode::new(
+                operator,
+                input_streams.clone().into_iter(),
+                self.clone(),
+                id,
+            );
+            let output_stream = node.output_stream();
+            for stream in input_streams.iter() {
+                self.connect_stream(stream, id, input_preference);
+            }
+            (node, output_stream)
+        })
```

Comment:
If these methods just immediately clone their inputs they should take owned values instead of references so we can let the caller decide when cloning needs to happen